lxd/network/network_utils.go

```diff
@@ -639,6 +639,111 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// noAvailableAddressErr is used by randomAddressInSubnet to indicate that the subnet was exhausted while searching for
+// IP addresses.
+type noAvailableAddressErr struct {
+	error
+}
+
+// Unwrap implements xerrors.Unwrap for noAvailableAddressErr.
+func (e noAvailableAddressErr) Unwrap() error {
+	return e.error
+}
+
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until all
+// IPs have been exhausted or the context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	ones, size := ipnet.Mask.Size()
+	subnetExponent := size - ones
+
+	// networkAddress is the first address in the subnet.
+	var networkAddress net.IP
+	if ipnet.IP.To4() != nil {
+		// Apply mask to get first address in subnet. Use To4 to ensure the IP is in the correct form for use with big.Int.
+		networkAddress = ipnet.IP.To4().Mask(ipnet.Mask)
+	} else {
+		networkAddress = ipnet.IP.To16().Mask(ipnet.Mask)
+	}
+
+	// If the subnet only has one address, or if the subnet is IPv4 and has two addresses return the network address.
+	if subnetExponent == 0 || (subnetExponent == 1 && ipnet.IP.To4() != nil) {
+		return networkAddress, nil
+	}
+
+	// If the subnet is IPv6 and has two addresses, return the second one (no broadcast address for IPv6).
+	if subnetExponent == 1 {
+		return big.NewInt(0).Add(big.NewInt(0).SetBytes(networkAddress), big.NewInt(1)).Bytes(), nil
```

Comment:
It shouldn't return it, but instead offer it to validate to check if its free right?